ext/MitsubaLoader/CMitsubaLoader.cpp

```diff
@@ -1369,7 +1405,25 @@ void instr_execute_BLEND(in instr_t instr, in uvec3 regs, in mat2 dUV, in bsdf_d
 void instr_execute(in instr_t instr, in uvec3 regs, in mat2 dUV, in vec3 L)
 {
 	bsdf_data_t bsdf_data = fetchBSDFDataForInstr(instr);
-	switch (instr.x & INSTR_OPCODE_MASK)
+	uint opcode = instr_getOpcode(instr);
+	uint ndf = instr_getNDF(instr);
+	bool twosided = instr_getTwosided(instr);
+	
+	//if (twosided)
+	//flip normal
+
+	vec2 a = getAlpha(opcode,ndf,instr,bsdf_data,dUV);
+	float a2 = a.x*a.y;
+	//TODO somehow get alpha here
+	float D = 0.0;
+	if (ndf==NDF_BECKMANN)
+		D = irr_glsl_beckmann(a2, currBSDFParams.isotropic.NdotH*currBSDFParams.isotropic.NdotH);
+	else if (ndf==NDF_GGX)
+		D = irr_glsl_ggx_trowbridge_reitz(a2, currBSDFParams.isotropic.NdotH*currBSDFParams.isotropic.NdotH);
+	else if (ndf==NDF_PHONG)
+		D = irr_glsl_blinn_phong(currBSDFParams.isotropic.NdotH, 1.0/a2);
+
+	switch (opcode)
```

Comment:
you could make a `getFresnelClass(opcode)` that returns `0,1,2` for None, Dielectric, and Conductor to compute the fresnel factor before the divergence. It could be predivided by `(4.0 * currBSDFParams.isotropic.interaction.NdotV)`

And then a `needsFresnelCorrection(opcode)` to tell us if the diffuse fresnel factors need to be computed as well

Reply:
I also notice that most BSDF (BRDF included) instructions need to fetch at least two textures (one roughness and other some parameter) and every texture has an associated scale... maybe we should fetch the textures without divergence?

Reply:
yea, but all parameters can come from texture OR constant and i'm unable to check which are available and which ones of them should be fetched from texture without being dependent on opcode. I guess it is possible to find common things in all instructions (at least BSDF/BRDF instructions), put `scale` always at the same byteoffset, put texture presence flag for 1st parameter always on bit N and this way i can always fetch at least this 1 parameter without divergence

Reply:
> I guess it is possible to find common things in all instructions (at least BSDF/BRDF instructions), put scale always at the same byteoffset, put texture presence flag for 1st parameter always on bit N and this way i can always fetch at least this 1 parameter without divergence

Sounds good